fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/KeyValueCursorBase.java

```diff
@@ -0,0 +1,410 @@
+            if (!stoppedReason.isPresent()) {
+                throw new RecordCoreException("limit manager stopped KeyValueCursor but did not report a reason");
+            }
+            nextResult = RecordCursorResult.withoutNextValue(continuationHelper(), stoppedReason.get());
+            return CompletableFuture.completedFuture(nextResult);
+        }
+    }
+
+    @Override
+    @Nonnull
+    public RecordCursorResult<K> getNext() {
+        return context.asyncToSync(FDBStoreTimer.Waits.WAIT_ADVANCE_CURSOR, onNext());
+    }
+
+    @Nonnull
+    private RecordCursorContinuation continuationHelper() {
+        return new Continuation(lastKey, prefixLength);
+    }
+
+    private static class Continuation implements RecordCursorContinuation {
+        @Nullable
+        private final byte[] lastKey;
+        private final int prefixLength;
+
+        public Continuation(@Nullable final byte[] lastKey, final int prefixLength) {
+            // Note that doing this without a full copy is dangerous if the array is ever mutated.
+            // Currently, this never happens and the only thing that changes is which array lastKey points to.
+            // However, if logic in KeyValueCursor or KeyValue changes, this could break continuations.
+            // To resolve it, we could resort to doing a full copy here, although that's somewhat expensive.
+            this.lastKey = lastKey;
+            this.prefixLength = prefixLength;
+        }
+
+        @Override
+        public boolean isEnd() {
+            return lastKey == null;
+        }
+
+        @Nonnull
+        @Override
+        public ByteString toByteString() {
+            if (lastKey == null) {
+                return ByteString.EMPTY;
+            }
+            ByteString base = ZeroCopyByteString.wrap(lastKey);
+            return base.substring(prefixLength, lastKey.length);
+        }
+
+        @Nullable
+        @Override
+        public byte[] toBytes() {
+            if (lastKey == null) {
+                return null;
+            }
+            return Arrays.copyOfRange(lastKey, prefixLength, lastKey.length);
+        }
+    }
+
+    /**
+     * A builder for {@link KeyValueCursorBase}.
+     * @param <T> the type of the concrete subclass of the builder
+     * This follows a pattern for nested builder class, hence is generic, so that subclasses can pass in their type (and
+     * implement {@link #getThis()}) to return the subclass builder correct type.
+     *
+     * <pre><code>
+     * KeyValueCursorSubclass.Builder.withSubspace(subspace)
+     *                     .setContext(context)
+     *                     .setRange(TupleRange.ALL)
+     *                     .setContinuation(null)
+     *                     .setScanProperties(ScanProperties.FORWARD_SCAN)
+     *                     .build()
+     * </code></pre>
+     */
+    @API(API.Status.MAINTAINED)
+    public abstract static class Builder<T extends Builder<T>> {
```

Comment:
Should this also be parameterized on the thing it builds, and then should there be an abstract `build` method that returns that thing? In theory, that would allow someone to write a method that took a `KeyValueCursorBase.Builder` and then it could build the cursor, but I'm not sure how useful that actually is.

Reply:
That's a good idea (™) but I couldn't make it work due to some strange quirks of the Java typing system. It appears that you can't add a second generic type marker to the Builder class for some reason (bug in Java interpreter?) and trying to achieve the same goal by introducing an abstract `build()` method on a different `<V extends KeyValue>` introduces `unchecked` errors.